File: ex309.java
// Create dp list buy and sell
// buy[i] = Math.max(buy[i - 1], sell[i - 2] - buy[i]), due to one day cooldown
// sell[i] = Math.max(sell[i - 1]. buy[i - 1] + sell[i])
// In this implementation, we use buy_1 = buy[i - 1], sell_1 = sell[i - 1], sell_2 = sell[i - 2]


class Solution {
    public int maxProfit(int[] prices) {
        // Special situation
        if (prices.length <= 1)    return 0;
        
        
        int buy = -prices[0], buy_1;
        int sell = 0, sell_1 = sell, sell_2;
        
        for (int i = 1; i < prices.length; ++i) {
            buy_1 = buy;
            sell_2 = sell_1;
            sell_1 = sell;
            
            buy = Math.max(buy_1, sell_2 - prices[i]);           
            sell = Math.max(sell_1, buy_1 + prices[i]);
        }
        
        return sell;
    }
}

File: code/ex309.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if (n == 0)    return 0;
        
        vector<int> hold(n), unHold(n);
        hold[0] = -prices[0];
        
        for (int i = 1; i < n; i++) {
            hold[i] = max(max(hold[i - 1], -prices[i]), i - 2 >= 0 ? unHold[i - 2] - prices[i] : INT_MIN);
            unHold[i] = max(unHold[i - 1], hold[i - 1] + prices[i]);
        }
        
        return unHold[n - 1];
    }
};